src/OrleansRuntime/GrainDirectory/LocalGrainDirectory.cs

```diff
@@ -469,175 +472,235 @@ public SiloAddress CalculateTargetSilo(GrainId grain, bool excludeThisSiloIfStop
-            if (owner.Equals(MyAddress))
+            if (forwardaddress == null)
             {
+                // we are the owner
                 UnregistrationsLocal.Increment();
-                // if I am the owner, remove the old activation locally
-                DirectoryPartition.RemoveActivation(addr.Grain, addr.Activation, force);
-                return TaskDone.Done;
+                var registrar = RegistrarManager.Instance.GetRegistrarForGrain(address.Grain);
+                await registrar.UnregisterAsync(address, force);
+            }
+            else
+            {
+                UnregistrationsRemoteSent.Increment();
+                // otherwise, notify the owner
+                await GetDirectoryReference(forwardaddress).UnregisterAsync(address, force, hopCount + 1);
             }
-            
-            UnregistrationsRemoteSent.Increment();
-            // otherwise, notify the owner
-            return GetDirectoryReference(owner).Unregister(addr, force, NUM_RETRIES);
         }
 
-        public Task UnregisterManyAsync(List<ActivationAddress> addresses)
+        private void AddToDictionary<K,V>(ref Dictionary<K, List<V>> dictionary, K key, V value)
         {
-            unregistrationsManyIssued.Increment();
-            return Task.WhenAll(
-                addresses.GroupBy(a => CalculateTargetSilo(a.Grain))
-                    .Select(g =>
+            if (dictionary == null) 
+               dictionary = new Dictionary<K,List<V>>();
+            List<V> list;
+            if (! dictionary.TryGetValue(key, out list))
+               dictionary[key] = list = new List<V>();
+            list.Add(value);
+        }
+
+        public async Task UnregisterManyAsync(List<ActivationAddress> addresses, int hopCount)
+        {
+            (hopCount > 0 ? UnregistrationsManyRemoteReceived : unregistrationsManyIssued).Increment();
+
+            Dictionary<SiloAddress, List<ActivationAddress>> forward = null;
+            var tasks = new List<Task>();
+
+            foreach (var address in addresses)
+            {
+                // see if the owner is somewhere else (returns null if we are owner)
+                var forwardAddress = this.CheckIfShouldForward(address.Grain, hopCount, "UnregisterManyAsync");
+
+                if (forwardAddress != null)
+                    AddToDictionary(ref forward, forwardAddress, address);
+
+                else
+                {
+                    // we are the owner
+                    UnregistrationsLocal.Increment();
+                    var registrar = RegistrarManager.Instance.GetRegistrarForGrain(address.Grain);
+                    tasks.Add(registrar.UnregisterAsync(address, true));
+                }
+            }
+
+            // on all silos other than first, we insert a retry delay and re-check destination
+            if (hopCount > 0 && forward != null)
+            {
+                await Task.Delay(RETRY_DELAY);
```

Comment:
Again, it's not clear to me why we need to delay here, especially when we are un-registering activations.